### PR TITLE
add chooseFromList with various options

### PR DESCRIPTION
**Pull request by orichters**
As discussed with @tscheypidi, a new generic function `chooseFromList` is added.

It is a cleaned version of the one I wrote in REMIND's [`start.R`](https://github.com/remindmodel/remind/blob/develop/start.R#L53-L133), based on various special versions such as `chooseFolder` that have been there for a while by different authors.

Features:
- works with character vectors and lists, and returns the same type
- `multiple` allows to select more than one 
- selection can be based on picking individual or multiple entries, including 2:5 which implies 2,3,4,5, `all` or regex matching
- the selected entries can be returned as a list, or as a boolean list with the same size as the original one
- supports selection based on groups supplied as vector/list names

@mikapfl: It would be great if you could check the lucode2::buildLibrary() changes and the tests.

### CODE REVIEW

File: R/chooseFromList.R [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#' chooseFromList
+#'
+#' Allows the user to select single or multiple elements from a list.
+#' Entries can be selected based on individual choice, groups, regex or all.
+#'
+#' @param theList list to be selected from
+#' @param group list with same dimension as theList with group names to allow to select whole groups
```
Comment (mikapfl) [inline R/chooseFromList.R:7]:
```suggestion
#' @param group list with same dimension as theList with group names to allow to select whole groups, optional
```

File: R/chooseFromList.R [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+#'
+#' @export
+chooseFromList <- function(theList, type = "runs", returnBoolean = FALSE, multiple = TRUE,
+                           allowEmpty = FALSE, group = FALSE, userinput = FALSE) {
+  originalList <- theList
+  booleanList <- rep(FALSE, length(originalList)) # set to FALSE
+  if (! isFALSE(group) && (isFALSE(multiple) || length(group) != length(originalList))) {
+    message("group must have same dimension as theList, or multiple not allowed. Group mode disabled")
+    group <- FALSE
+  }
```
Comment (mikapfl) [inline R/chooseFromList.R:29]: I think it is better to be specific with the error message, so make two `ifs`. Also, group with a different dimension than theList is simply an error, not something I would try to recover.

File: R/chooseFromList.R [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  m <- paste0("\n\nPlease choose ", type, ":\n\n")
+  # in group mode, paste group after each entry and add them to theList as options
+  if (! isFALSE(group)) {
```
Comment (mikapfl) [inline R/chooseFromList.R:32]: just double-checking: `! isFALSE(group)` is not the same as `isTRUE(group)` or `boolean(group)` or some other formulation which is saner?

File: R/chooseFromList.R [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  }
+  # go through list of userinput, and turn 2:5 into 2,3,4,5 etc.
+  tmp <- NULL
+  for (i in seq_along(identifier)) {
+    if (length(strsplit(identifier, ":")[[i]]) > 1) {
+      tmp <- c(tmp, as.numeric(strsplit(identifier, ":")[[i]])[1]:as.numeric(strsplit(identifier, ":")[[i]])[2])
+    } else {
+      tmp <- c(tmp, as.numeric(identifier[i]))
+    }
+  }
```
Comment (mikapfl) [inline R/chooseFromList.R]: this could well be extracted into its own (private) function in the same file. Then it could also be tested with one or two quick tests. (optional, if you like to do some refactoring to get the cyclomatic complexity of the function down)

**Review by mikapfl — CHANGES_REQUESTED** [review]
Looks good to me! In particular, the tests inspire confidence that at least the happy path works well.

The function is quite long and therefore at times hard to follow. Maybe it would be worth to try to split it into three or four functions (in the same file), all called by the main function in order? Like `printOptions()`, `parseRawInput()`, `parseRegex()` or so? Would certainly help whoever has to come back to this in 3-4 years. But I'd be also fine with merging as-is if you don't have the time for refactoring.

**Comment by orichters** [comment]
I don't understand why the checks fail: https://github.com/pik-piam/gms/runs/7728861581?check_suite_focus=true
It does not happen while running the checks on the cluster, seems to be related to [some R mode change](https://stackoverflow.com/questions/72848442/r-warning-lengthx-2-1-in-coercion-to-logical1) and I'm unable to debug it. Maybe @mikapfl you have an idea?

File: R/chooseFromList.R (hunk 3 above)
Comment (orichters) [reply R/chooseFromList.R:32]: no, because if `group <- c("A", "B")`, then `isTRUE(group)` is FALSE. I want `TRUE` only if `group` is indeed `FALSE`, nothing else.

**Comment by mikapfl** [comment]
> I don't understand why the checks fail: https://github.com/pik-piam/gms/runs/7728861581?check_suite_focus=true It does not happen while running the checks on the cluster, seems to be related to [some R mode change](https://stackoverflow.com/questions/72848442/r-warning-lengthx-2-1-in-coercion-to-logical1) and I'm unable to debug it. Maybe @mikapfl you have an idea?

I get these errors also when testing on my laptop with R 4.2, I guess you don't get them on the cluster because there we have R 4.1.

I don't know immediately what is happening there and why it is unhappy.

**Comment by mikapfl** [comment]
I guess it would be easier to debug if R could tell me at which line the coercion to logical doesn't work. I'll never get used to these utterly incomplete tracebacks from R.

File: R/chooseFromList.R [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    message("Try again, not all in list: ", paste(identifier, collapse = ", "))
+    return(chooseFromList(originalList, type, returnBoolean, multiple, allowEmpty, group))
+  }
+  # interpret group inputs
+  if (! isFALSE(group)) {
+    selectedGroups <- sub("^Group: ", "", theList[intersect(identifier, groupsids)])
+    identifier <- unique(c(identifier[! identifier %in% groupsids], which(group %in% selectedGroups) + 1))
+  }
+  # if search by pattern is selected, ask for pattern and interpret it
+  if (multiple && length(identifier == 1) && identifier == length(theList)) {
```
Comment (mikapfl) [inline R/chooseFromList.R:82]: here, `identifier` can be a list with more than one entry. therefore `identifier == length(theList)` can also be a list with mor than one entry. Coercing it to boolean will then raise a warning.

Comment (mikapfl) [reply R/chooseFromList.R:82]:
```suggestion
  if (multiple && length(identifier) == 1 && identifier == length(theList)) {
```
Is probably more correct?

**Comment by orichters** [comment]
> I guess it would be easier to debug if R could tell me at which line the coercion to logical doesn't work. I'll never get used to these utterly incomplete tracebacks from R.

Yeah, terrible! But there are only two `length(...) > 1` statements, [here](https://github.com/pik-piam/gms/pull/32/files#diff-228cc631ecba462ee423b90b5a3a4b6d59fe2b3d309c7dfde8f34781a88c7d73R60) and [here](https://github.com/pik-piam/gms/pull/32/files#diff-228cc631ecba462ee423b90b5a3a4b6d59fe2b3d309c7dfde8f34781a88c7d73R68), but I already added the `isTRUE` such that it coerces to a unique result. Maybe `all()` instead?

File: R/chooseFromList.R (hunk 5 above)
Comment (orichters) [reply R/chooseFromList.R:82]: Baaah, Must be `length(identifier) == 1` of course :(

Comment (mikapfl) [reply R/chooseFromList.R:82]: I got to the right line with `options(warn=2, error=recover)` - seems to be the go-to debugging tool for when you need to know where the warning actually happened.

File: R/chooseFromList.R [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  # go through list of userinput, and turn 2:5 into 2,3,4,5 etc.
+  tmp <- NULL
+  for (i in seq_along(identifier)) {
+    if (isTRUE(length(strsplit(identifier, ":")[[i]]) > 1)) {
+      tmp <- c(tmp, as.numeric(strsplit(identifier, ":")[[i]])[1]:as.numeric(strsplit(identifier, ":")[[i]])[2])
+    } else {
+      tmp <- c(tmp, as.numeric(identifier[i]))
+    }
+  }
+  identifier <- tmp
```
Comment (chroetz) [inline R/chooseFromList.R:66]: Just a note: This is a short version for parsing the user input: `eval(parse(text = paste("c(", input, ")")))`.
It puts the input into `c()` and evaluates it as an R expression.

**Comment by chroetz** [comment]
Thanks for doing this, @orichters! It also bothered me for some time that there a several duplications of functions like this in different scripts in the REMIND repo. I even added my [own](https://github.com/remindmodel/remind/blob/050cf6fa622ee8eab3f7dd51f8d1e042764d0d42/scripts/output/comparison/compareScenarios2.R#L31)... In that implementation, there are default choices which are selected if the user just hits enter. Furthermore, the default choices are colored with `crayon`. What do you think about adding this feature to `gms::chooseFromList()`?

File: R/chooseFromList.R (hunk 2 above)
Comment (orichters) [reply R/chooseFromList.R]: done

File: R/chooseFromList.R (hunk 1 above)
Comment (orichters) [reply R/chooseFromList.R]: done

File: R/chooseFromList.R (hunk 6 above)
Comment (orichters) [reply R/chooseFromList.R:69]: I'm not that sure this is a good idea, at least without extensive checks for the input the user provides. This is at least what I learned from [Little Bobby Tables](https://xkcd.com/327/).

Comment (mikapfl) [reply R/chooseFromList.R:69]: The user is running arbitrary code already. There is no security boundary here. If the user wants to crash the program, they can just hit `ctrl+c`. So feel free to `eval` whatever the user hands to you if it makes your code cleaner.

The problem I see with `eval`'ing here is more about error messages. If the user inputs "`list(1,2,3)`", that could lead to absolutely hilarious error messages down the line. Here, it is much more helpful to tell the user early that "you must input a number".

File: R/chooseFromList.R [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+#' @param allowEmpty TRUE: allows you not to select anything (returns NA). FALSE: must select something
+#' @param type string to be shown to user to understand what they choose
+#' @param userinput string provided by the user. If not supplied, user is asked
+#'
+#' @return either a boolean list with same length as input or only the selected list items.
+#'
+#' @author Oliver Richters
+#'
+#' @importFrom stringr str_pad
+#'
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/chooseFromList.R:20]: This function needs examples.

File: R/chooseFromList.R [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  }
+  m <- paste0("\n\nPlease choose ", type, ":\n\n")
+  # in group mode, paste group after each entry and add them to theList as options
+  if (! isFALSE(group)) {
+    groups <- sort(unique(group))
+    groupsids <- seq(length(originalList) + 2, length(originalList) + length(groups) + 1)
+    theList <- c(paste0(str_pad(theList, max(nchar(originalList), 6), side = "right"), " ", group),
+                 paste("Group:", groups))
+    m <- c(m, paste0(str_pad("", max(nchar(originalList), 6) + nchar(length(theList) + 2) + 2, side = "right"),
+              " Group"))
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/chooseFromList.R:41]: You are missing a new line here, `Group` gets conflated with `1: all` in the output:
```
> chooseFromList(theList = list('one', 'two', 'three'), group = c('one' = 'odd', 'two' = 'even', 'three' = 'odd'))


Please choose runs:

          Group1: all
2: one    odd
3: two    even
4: three  odd
5: Group: even
6: Group: odd
7: Search by pattern...
Numbers entered as 2,4:6,9 (runs): 
```

File: R/chooseFromList.R [hunk 9]
```diff
@@ -0,0 +1,109 @@
+#' chooseFromList
+#'
+#' Allows the user to select single or multiple elements from a list.
+#' Entries can be selected based on individual choice, groups, regex or all.
+#'
+#' @param theList list to be selected from
+#' @param group list with same dimension as theList with group names to allow to select whole groups, optional
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/chooseFromList.R:7]: You call it a "list", but the function breaks when an actual `list` is supplied:
```
> chooseFromList(theList = list('one', 'two', 'three'), group = list('one' = 'odd', 'two' = 'even', 'three' = 'odd'))
Error in sort.int(x, na.last = na.last, decreasing = decreasing, ...) : 
  'x' must be atomic
```
Either fix the code, or call it a (character) vector.

Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/chooseFromList.R:7]: Why are `theList` and `group` separate parameters?  Users have to repeat the same data, once as data proper (`theList`), once as names to a named vector.  Quite cumbersome.
Why not use just one input, with list items to choose as data and optional groups as names?

File: R/chooseFromList.R [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+#' @param returnBoolean TRUE: returns list with dimension of theList with FALSE or TRUE
+#'                      FALSE: returns selected entries of theList
+#' @param multiple TRUE: allows to select multiple entries. FALSE: no
+#' @param allowEmpty TRUE: allows you not to select anything (returns NA). FALSE: must select something
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/chooseFromList.R:11]: Users should always have an explicit option to choose nothing and abort.  Else they just hit CTRL-C.
If there's too much legacy code that can't handle the user choosing nothing, this could be changed to a parameter that quits the R script if the user chooses nothing, instead of returning control (probably the most common use case for us).

File: R/chooseFromList.R [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+#' @param type string to be shown to user to understand what they choose
+#' @param userinput string provided by the user. If not supplied, user is asked
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/chooseFromList.R:13]: Out of curiosity: what is the use case for this?

**Review by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q — CHANGES_REQUESTED** [review]

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
Also, have you given thought to improving list presentation?  The [list in `choose_slurmConfig.R`](https://github.com/remindmodel/remind/blob/050cf6fa622ee8eab3f7dd51f8d1e042764d0d42/scripts/start/choose_slurmConfig.R#L16-L35) has usability features (section delimiters) that could improve lots of other use cases.  They could be implemented as a special case/escape sequence like treating all items starting with five dashes as separators
```
> modes <- c("SLURM standby               12   nash H12             [recommended]",
           "SLURM standby               13   nash H12 coupled",
           "SLURM standby               16   nash H12+",
           "SLURM standby                1   nash debug, testOneRegi, quick",
           "-----",
           "SLURM priority              12   nash H12             [recommended]",
           "SLURM priority              13   nash H12 coupled",
           "SLURM priority              16   nash H12+", 
           "SLURM priority               1   nash debug, testOneRegi, quick", 
           "-----",
           "SLURM short                 12   nash H12",
           "SLURM short                 16   nash H12+", 
           "SLURM short                  1   nash debug, testOneRegi, quick", 
           "SLURM medium                 1   negishi",
           "SLURM long                   1   negishi", 
           "-----",
           "SLURM medium                12   nash - long calibration", 
           "SLURM medium                16   nash - long calibration",
           "-----", 
           "direct, without SLURM")
> chooseFromList(modes, multiple = FALSE)


Please choose runs:

 1: SLURM standby               12   nash H12             [recommended]
 2: SLURM standby               13   nash H12 coupled
 3: SLURM standby               16   nash H12+
 4: SLURM standby                1   nash debug, testOneRegi, quick
    -------------------------------------------------------------------
 5: SLURM priority              12   nash H12             [recommended]
 6: SLURM priority              13   nash H12 coupled
 7: SLURM priority              16   nash H12+
 8: SLURM priority               1   nash debug, testOneRegi, quick
    -------------------------------------------------------------------
 9: SLURM short                 12   nash H12
10: SLURM short                 16   nash H12+
11: SLURM short                  1   nash debug, testOneRegi, quick
12: SLURM medium                 1   negishi
13: SLURM long                   1   negishi
    -------------------------------------------------------------------
14: SLURM medium                12   nash - long calibration
15: SLURM medium                16   nash - long calibration
    -------------------------------------------------------------------
16: direct, without SLURM
Number (runs):
```
Maybe for future improvement `;)`

File: R/chooseFromList.R (hunk 9 above)
Comment (orichters) [reply R/chooseFromList.R:7]: `group` is optional, see last word in the param description. I like your idea though to merge it into one list…

File: R/chooseFromList.R (hunk 11 above)
Comment (orichters) [reply R/chooseFromList.R:13]: Main reason: for `testthat` for [the function tests](https://github.com/pik-piam/gms/pull/32/files#diff-5b23d4fd60a302196db0d52c010ef6dfac47ef663a0ab602ba8007bf8c594600), because there seems to be no easy solution to deal with the `getLine()` calls, and I agreed with @mikapfl that this is a good option.

Minor reason: If at some point somebody want to migrate `choose_slurmConfig()` to this option, it allows to specify the identifier aka userinput as input from a scenario config file.

File: R/chooseFromList.R (hunk 9 above)
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [reply R/chooseFromList.R:7]: That's why I suggested group names to be optional, see third to last word in my suggestion.

Comment (orichters) [reply R/chooseFromList.R:7]: I think this is a misunderstanding. You essentially had `theList <- c(1, 2, "A")` and you optionally `group <- c("Num", "Num", "Letter")`. But I prefer it with `theList <- c(Num = 1, Num = 2, Letter = "A")` as you suggested.

File: R/chooseFromList.R (hunk 10 above)
Comment (orichters) [reply R/chooseFromList.R]: Ok. I deleted the `allowEmpty` option. An empty list or character vector is then returned.

File: R/chooseFromList.R (hunk 9 above)
Comment (orichters) [reply R/chooseFromList.R]: I implemented the `group` feature as list / character vector names.

Comment (orichters) [reply R/chooseFromList.R]: The problem was that you did not understand the `group` feature as I implemented it. But as I re-implemented it, this should be fine. `chooseFromList(theList = list("odd" = 1, "even" = "two", "odd" = "three"))` works.

File: R/chooseFromList.R (hunk 7 above)
Comment (orichters) [reply R/chooseFromList.R:29]: done

**Comment by orichters** [comment]
> Thanks for doing this, @orichters! It also bothered me for some time that there a several duplications of functions like this in different scripts in the REMIND repo. I even added my [own](https://github.com/remindmodel/remind/blob/050cf6fa622ee8eab3f7dd51f8d1e042764d0d42/scripts/output/comparison/compareScenarios2.R#L31)... In that implementation, there are default choices which are selected if the user just hits enter. Furthermore, the default choices are colored with `crayon`. What do you think about adding this feature to `gms::chooseFromList()`?

I added the option to specify a group named `default` which is selected if the user types nothing. I tried a bit with coloring, but the characters added by the crayon function messes with the layout of the groups, so people have to live without colors.

File: R/chooseFromList.R (hunk 4 above)
Comment (orichters) [reply R/chooseFromList.R]: I used the quicker version by @chroetz and put the pattern matching into a single function, so R is not annoyed about cyclomatic complexity anymore.

**Review by mikapfl — APPROVED** [review]

**Comment by orichters** [comment]
> Also, have you given thought to improving list presentation? The [list in `choose_slurmConfig.R`](https://github.com/remindmodel/remind/blob/050cf6fa622ee8eab3f7dd51f8d1e042764d0d42/scripts/start/choose_slurmConfig.R#L16-L35) has usability features (section delimiters) that could improve lots of other use cases. They could be implemented as a special case/escape sequence like treating all items starting with five dashes as separators. Maybe for future improvement `;)`

The other necessity before slurmConfig can be chosen based on this function would be to implement that the options presented to the user are different to what is actually selected…

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
> The other necessity before slurmConfig can be chosen based on this function would be to implement that the options presented to the user are different to what is actually selected…

I don't follow.  How so?

**Comment by orichters** [comment]
> > The other necessity before slurmConfig can be chosen based on this function would be to implement that the options presented to the user are different to what is actually selected…
> 
> I don't follow. How so?

The users gets to select based on
```
 1: SLURM standby               12   nash H12             [recommended]
...
```
but the actual selection the function returns is
```
--qos=standby --nodes=1 --tasks-per-node=12
...
```
`chooseFromList` does not (yet) allow this.

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
> The users gets to select based on
> 
> ```
>  1: SLURM standby               12   nash H12             [recommended]
> ...
> ```
> 
> but the actual selection the function returns is
> 
> ```
> --qos=standby --nodes=1 --tasks-per-node=12
> ...
> ```
> 
> `chooseFromList` does not (yet) allow this.

No business of `chooseFromList()`.  Simply pass
```
which(chooseFromList(..., returnBoolean = TRUE))
```
into the [following `switch()` statement](https://github.com/remindmodel/remind/blob/050cf6fa622ee8eab3f7dd51f8d1e042764d0d42/scripts/start/choose_slurmConfig.R#L50-L66).

**Review by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q — APPROVED** [review]

File: R/chooseFromList.R (hunk 6 above)
Comment (orichters) [reply R/chooseFromList.R]: I used two checks now: It checks whether the whole user-input consists of `[0-9,:]` and whether the `eval` produces an error. That should catch the most quirky user input.
